programs/bpf_loader/src/syscalls/mem_ops.rs

```diff
@@ -152,16 +121,653 @@ declare_syscall!(
     ) -> Result<u64, EbpfError> {
         mem_op_consume(invoke_context, n)?;
 
-        let s = translate_slice_mut::<u8>(
+        if invoke_context
+            .feature_set
+            .is_active(&feature_set::bpf_account_data_direct_mapping::id())
+        {
+            memset_non_contiguous(dst_addr, c as u8, n, memory_mapping)
+        } else {
+            let s = translate_slice_mut::<u8>(
+                memory_mapping,
+                dst_addr,
+                n,
+                invoke_context.get_check_aligned(),
+                invoke_context.get_check_size(),
+            )?;
+            s.fill(c as u8);
+            Ok(0)
+        }
+    }
+);
+
+fn memmove(
+    invoke_context: &mut InvokeContext,
+    dst_addr: u64,
+    src_addr: u64,
+    n: u64,
+    memory_mapping: &MemoryMapping,
+) -> Result<u64, EbpfError> {
+    if invoke_context
+        .feature_set
+        .is_active(&feature_set::bpf_account_data_direct_mapping::id())
+    {
+        memmove_non_contiguous(dst_addr, src_addr, n, memory_mapping)
+    } else {
+        let dst_ptr = translate_slice_mut::<u8>(
             memory_mapping,
-            s_addr,
+            dst_addr,
             n,
             invoke_context.get_check_aligned(),
             invoke_context.get_check_size(),
-        )?;
-        for val in s.iter_mut().take(n as usize) {
-            *val = c as u8;
-        }
+        )?
+        .as_mut_ptr();
+        let src_ptr = translate_slice::<u8>(
+            memory_mapping,
+            src_addr,
+            n,
+            invoke_context.get_check_aligned(),
+            invoke_context.get_check_size(),
+        )?
+        .as_ptr();
+
+        unsafe { std::ptr::copy(src_ptr, dst_ptr, n as usize) };
         Ok(0)
     }
-);
+}
+
+fn memmove_non_contiguous(
+    dst_addr: u64,
+    src_addr: u64,
+    n: u64,
+    memory_mapping: &MemoryMapping,
+) -> Result<u64, EbpfError> {
+    let reverse = dst_addr.wrapping_sub(src_addr) < n;
+    iter_memory_pair_chunks(
+        AccessType::Load,
+        src_addr,
+        AccessType::Store,
+        dst_addr,
+        n,
+        memory_mapping,
+        reverse,
+        |src_host_addr, dst_host_addr, chunk_len| {
+            unsafe {
+                std::ptr::copy(
+                    src_host_addr as *const u8,
+                    dst_host_addr as *mut u8,
+                    chunk_len,
+                )
+            };
+            Ok(0)
+        },
+    )
+}
+
+// Marked unsafe since it assumes that the slices are at least `n` bytes long.
+unsafe fn memcmp(s1: &[u8], s2: &[u8], n: usize) -> i32 {
+    for i in 0..n {
+        let a = *s1.get_unchecked(i);
+        let b = *s2.get_unchecked(i);
+        if a != b {
+            return (a as i32).saturating_sub(b as i32);
+        };
+    }
+
+    0
+}
```

Comment:
This syscall can be used manually but it's also implicitly used by rustc. I think that memcmp must return `< 0`, `0` or `> 0` for lt, eq, gt, but it isn't clear whether the absolute value in itself is part of the API. All memcmp implementations I've ever seen do return the difference between two bytes, so I'm not inclined to try and innovate :P 

But in any case, the existing code exits early in case of diff, while the `sum()` version always scans the whole length.